symengine/dict.h

```diff
@@ -185,6 +185,165 @@ std::ostream &operator<<(std::ostream &out, const SymEngine::vec_basic &d);
+        }
+    }
+
+    ODictWrapper(const Value &p)
+    {
+        if (p != Value(0))
+            dict_[0] = p;
+    }
+
+    Wrapper &operator=(Wrapper &&other) SYMENGINE_NOEXCEPT
+    {
+        if (this != &other)
+            dict_ = std::move(other.dict_);
+        return static_cast<Wrapper &>(*this);
+    }
+
+    friend Wrapper operator+(const Wrapper &a, const Wrapper &b)
+    {
+        Wrapper c = a;
+        c += b;
+        return c;
+    }
+
+    Wrapper &operator+=(const Wrapper &other)
+    {
+        for (auto &iter : other.dict_) {
+            auto t = dict_.lower_bound(iter.first);
+            if (t != dict_.end() and t->first == iter.first) {
+                t->second += iter.second;
+                if (t->second == 0) {
+                    dict_.erase(t);
+                }
+            } else {
+                dict_.insert(t, {iter.first, iter.second});
+            }
+        }
+        return static_cast<Wrapper &>(*this);
+    }
+
+    friend Wrapper operator-(const Wrapper &a, const Wrapper &b)
+    {
+        Wrapper c = a;
+        c -= b;
+        return c;
+    }
+
+    Wrapper operator-() const
+    {
+        ODictWrapper c = *this;
+        for (auto &iter : c.dict_)
+            iter.second *= -1;
+        return static_cast<Wrapper &>(c);
+    }
+
+    Wrapper &operator-=(const Wrapper &other)
+    {
+        for (auto &iter : other.dict_) {
+            auto t = dict_.lower_bound(iter.first);
+            if (t != dict_.end() and t->first == iter.first) {
+                t->second -= iter.second;
+                if (t->second == 0) {
+                    dict_.erase(t);
+                }
+            } else {
+                dict_.insert(t, {iter.first, -iter.second});
+            }
+        }
+        return static_cast<Wrapper &>(*this);
+    }
+
+    friend Wrapper operator*(const Wrapper &a, const Wrapper &b)
+    {
+        Wrapper c = a;
+        c *= b;
+        return c;
+    }
+
+    Wrapper &operator*=(const Wrapper &other)
+    {
+        // are these checks really needed? they'll be handled
+        // automatically
+
+        // if (dict_.empty())
+        //     return static_cast<Wrapper &>(*this);
+
+        // if (other.dict_.empty()) {
+        //     *this = other;
+        //     return static_cast<Wrapper &>(*this);
+        // }
+
+        // //! other is a just constant term
+        // if (other.dict_.size() == 1
+        //     and other.dict_.find(0) != other.dict_.end()) {
+        //     for (const auto &i1 : dict_)
+        //         for (const auto &i2 : other.dict_)
+        //             dict_[i1.first + i2.first] = i1.second * i2.second;
+        //     return static_cast<Wrapper &>(*this);
+        // }
+
+        std::map<Key, Value> p;
+        for (const auto &i1 : dict_)
+            for (const auto &i2 : other.dict_)
+                p[i1.first + i2.first] += i1.second * i2.second;
+
+        *this = Wrapper(p);
```

Comment:
You are right. This would call the assignment operator.
To avoid the extra allocation in `Wrapper(p)`, you can canonicalize the dict `p` and swap `p` and `this->dict_`